src/api/schema/components/source.rs

```diff
@@ -70,3 +83,99 @@ impl Source {
         metrics::by_component_name(&self.0.name).to_source_metrics(&self.0.component_type)
     }
 }
+
+#[derive(Default, InputObject)]
+pub struct SourcesFilter {
+    name: Option<Vec<filter::StringFilter>>,
+    component_type: Option<Vec<filter::StringFilter>>,
+    output_type: Option<Vec<filter::EqualityFilter<SourceOutputType>>>,
+    or: Option<Vec<Self>>,
+}
+
+impl filter::CustomFilter<Source> for SourcesFilter {
+    fn matches(&self, source: &Source) -> bool {
+        filter_check!(
+            self.name
+                .as_ref()
+                .map(|f| f.iter().all(|f| f.filter_value(source.0.get_name()))),
+            self.component_type.as_ref().map(|f| f
+                .iter()
+                .all(|f| f.filter_value(source.0.get_component_type()))),
+            self.output_type.as_ref().map(|f| f
+                .iter()
+                .all(|f| f.filter_value(source.0.output_type.into())))
+        );
+        true
+    }
+
+    fn or(&self) -> Option<&Vec<Self>> {
+        self.or.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::config::DataType;
+
+    /// Generate component fixes for use with tests
+    fn source_fixtures() -> Vec<Source> {
+        vec![
+            Source(Data {
+                name: "gen1".to_string(),
+                component_type: "generator".to_string(),
+                output_type: DataType::Any,
+            }),
+            Source(Data {
+                name: "gen2".to_string(),
+                component_type: "generator".to_string(),
+                output_type: DataType::Log,
+            }),
+            Source(Data {
+                name: "gen3".to_string(),
+                component_type: "generator".to_string(),
+                output_type: DataType::Metric,
+            }),
+        ]
+    }
+
+    #[test]
+    fn filter_output_type() {
+        struct Test {
```

Comment:
:+1: I forgot you can define structs in whatever scope. I like this, it makes the parameters more clear to me than our usual use of tuples for these table-driven tests.